### PR TITLE
Design document: integrating numpyro_forecast with dynestyx

### PR DESCRIPTION
**Pull request by juanitorduz**
Part of #34 (see also BasisResearch/dynestyx#264).

Adds `DYNESTYX_INTEGRATION_DESIGN.md` at the repo root: a detailed scaffolding of how `numpyro_forecast` and `dynestyx` compose, replacing the earlier draft on the `dynestyx` branch (which targeted the pre-refactor class API and was blocked by an ArviZ conflict that dynestyx v0.5.0 resolved). The second revision incorporates the dynestyx maintainers' review (comment on this PR).

## The design in one paragraph

A dynestyx state space model is written as a plain `numpyro_forecast` model `(covariates, data=None)`. One new model building block, `state_space_series(h, name, y, dynamics, conditioner=...)`, takes the dynestyx handler that conditions the observed window as its `conditioner`: `Filter` or `Smoother` (marginal likelihood as a NumPyro factor, latent path integrated out) or `LatentPathBuilder` (explicit path), so marginalized and explicit-path inference differ by one argument. The block adds the horizon bookkeeping: `Simulator` outside the conditioner for the rollout (registered as `"forecast"`), and, when a driver calls the model without data, a draw of the in-window states from the conditioner's posterior over the path plus one observation per step (registered as `"obs"`), which is what `predict_in_sample` and `to_datatree` read. The observed window travels in `covariates`, as for `ssoe`. Every nf driver, `backtest` with `eval_train=True` included, works unchanged.

## What changed after the review

- In-sample prediction through the `Smoother` (or the explicit path): `predict_in_sample`, `to_datatree` and `backtest(eval_train=True)` now apply; the smoother costs the same per gradient as the filter (0.17 ms, identical log density and NUTS draws).
- `LatentPathBuilder` as a third strategy, with its jit requirements pinned (jax time grids, one shared instance so its cached observation layout is available under `jit`); the exact-observation idiom (`DiracIdentityObservation`) verified on an AR(1); Section 9's SSOE claim narrowed to the innovations-form ETS family.
- Explicit, possibly irregular, `times` instead of a scalar `dt` (verified pass-through of an irregular grid).
- `LTI_discrete(..., D=...)` infers `control_dim` since dynestyx#361; the explicit `DynamicalModel` construction is gone from the examples.

Every dynestyx fact the block relies on was verified with throwaway probes (Appendix A, now A.1 to A.10), against v0.5.0 and against the `ml-bugfix-controldim-LTI` branch.

## Scope

Document only; no code changes. The example notebook implementing the design is the stacked PR #126.

### CODE REVIEW

**Comment by juanitorduz** [comment]
@claude review

**Comment by mattlevine22** [comment]
@juanitorduz Thanks for making this design doc, it looks great!

Here are the few things on noticed from wearing my Dynestyx hat:

1. `predict_in_sample` not implemented---perhaps have a look at our [Smoothers Tutorial](https://basisresearch.github.io/dynestyx/stable/tutorials/gentle_intro/09_discrete_smoothing/)?
- If you want the full smoothing distribution, this should be what you want.
- If, instead, you want to use some (but not all) of your observed data and smooth between them...then indeed our smoothers are not yet supporting this (mostly due to some [outstanding issues with unified indexing](https://github.com/BasisResearch/dynestyx/issues/272))

2. You may want to build a 3rd model in your demos that use `dsx.LatentPathBuilder` instead of `dsx.Filter`:
- This would use Dynestyx's internal machinery to explicitly create latent states (similar to `nf.innovations`)
- Why?
    - less code to distinguish between DO/DONT marginalize configurations
    - when paired with `dsx.Discretizer`, it allows you to explicitly create latent states for (discretized) continuous-time SDE/ODEs (this way you wouldn't have to manage numerical solvers in `nf.innovations` and can adhere to separation-of-concerns...in large part, the continuous time setting is the one I believe Dynestyx offers the most distinct support for). This is of course a future direction, but another motivating reason to try out `dsx.LatentPathBuilder` sooner rather than later.
    - when paired with `observation_model=dsx.DiracIdentityObservation()`, it will allow you to write the single-source-of-error models which Sec. 9 says are incompatible with Dynestyx. I recognize the `Identity` idiom may be a bit silly, but it works well and is rather clear haha.
    - it also has some nice features around handling NaN missingness

3. The `filtered_series` API for `dt`
- The choice of a single `dt` here runs against our preferences to allow users to explicitly state (possibly irregular) `times` in Dynestyx, where we have made special efforts to connect irregularly spaced data to continuous-time models.
- Overall, this is more up to you and how you envision usage; I suppose it matters more when you have continuous-time models, but could even show up with time-dependent parameters (where they depend on "real-time", not index time).

4. At the end of Sec. 6, you needed to make a special `local_level_regression` using `dsx.DynamicalModel` instead of `dsx.LTIDiscrete`:
- This was a bug in Dynestyx where it only infers the control dimension using `B` and doesn't check `D`. I put a [PR for this](https://github.com/BasisResearch/dynestyx/pull/361), so that should get fixed soon. Thanks for finding that!

**Comment by juanitorduz** [comment]
Thanks @mattlevine22 ! I addressed your changes and propagated them into the example notebook https://github.com/juanitorduz/numpyro_forecast/pull/126 ... any feedback on the API? For the moment is quite open :)

File: DYNESTYX_INTEGRATION_DESIGN.md [hunk 1]
```diff
@@ -1,411 +1,539 @@
...
+2. Fitting always happens with `future == 0` (nf's invariant), so the simulator and its `_predicted_*` sites never appear in the guide or in the MCMC trace: the forecast branch of the block is only ever executed under `Predictive`. The path site of the `LatentPathBuilder` has the shape of the observed window whatever the horizon, so posterior substitution works for it exactly as the `_future` suffix makes it work for `innovations` and `markov_series`.
+3. Under `Predictive(posterior_samples=...)` the model re-conditions on the window for every posterior draw and then rolls the conditioned state forward, so the forecast is conditioned on the data through the state, with the draw's own parameters, which is the correct posterior predictive $\int p(y_{T+1:T+H} \mid x_T, \theta) \, p(x_T \mid y_{1:T}, \theta) \, p(\theta \mid y_{1:T}) \, dx_T \, d\theta$. The in-sample predictive is the same construction over the window: $\int p(y_t \mid x_t, \theta) \, p(x_t \mid y_{1:T}, \theta) \, p(\theta \mid y_{1:T}) \, dx_t \, d\theta$, per step, which is what nf's direct models produce marginally when `predict_in_sample` replays their sampled path.
```
Comment (mattlevine22) [inline DYNESTYX_INTEGRATION_DESIGN.md]: This is right.

If this "re-conditioning" becomes prohibitively expensive, you could do cache the final filtered/smoothed state `filtered_state_means[-1]` associated with each posterior parameter sample, and then produce forecasted rollouts using `dsx.simulate` (vmapped over pairs of parameter + initial_state_dist).

Note that the reason we allow for this re-conditioning is so that, post-posterior-sampling, a user could experiment with different data streams and/or control interventions. I recognize, though, that in many cases this is just wasted computation; it would be nice for us to offer some sort of "replay" or "frozen_data=True" feature in Dynestyx.

File: DYNESTYX_INTEGRATION_DESIGN.md [hunk 2]
```diff
@@ -1,411 +1,539 @@
...
...el building block, `state_space_series(h, name, y, dynamics, conditioner=...)`. The `conditioner` is the `dynestyx` handler that interprets `dsx.sample` over the observed window, which is how `dynestyx` separates the model from its inference: a `Filter` or a `Smoother` marginalizes the latent path and adds the marginal log likelihood $\log p(y_{1:T} \mid \theta)$ as a NumPyro factor, a `LatentPathBuilder` samples the path explicitly. The block adds nf's horizon bookkeeping on top of whichever conditioner is passed: while forecasting it nests the conditioner inside a `Simulator`, whose posterior rollout starts from the conditioned state at the last observed step, and returns the horizon draws that the model registers as `"forecast"`; when a driver calls the model without data it draws the in-window states from the conditioner's posterior over the path (the smoothing distribution, or the explicit path) and one observation per step, which is what `predict_in_sample` and `to_datatree` read as `"obs"`. Nothing else changes: `SVI`, `MCMC`, `forecast`, `predict_in_sample`, `to_datatree`, `backtest` (including `eval_train`), the metrics and the ArviZ export all work on the model unchanged.
```
Comment (mattlevine22) [inline DYNESTYX_INTEGRATION_DESIGN.md]: By the way, is `predict_in_sample` a "smoothing" task or is it more like a prior-predictive task?

Just to be clear, the goal of a smoother is to estimate `p(x1...xT | y1...yT, theta)`.
If `data=None`, I'm not sure what value the Smoother is bringing...

The smoothers do also compute marginal likelihood (and sometimes do a better job), so I'm very glad you've included the smoothers regardless of how to support `predict_in_sample`.

File: DYNESTYX_INTEGRATION_DESIGN.md [hunk 3]
```diff
@@ -1,411 +1,539 @@
...
-- `controls` are the exogenous inputs over the full horizon, shape `(duration, control)`, normally the model's `covariates` themselves. They become `ctrl_values` on the grid `ctrl_times = arange(duration) * dt`, which covers both the observation times and the prediction times, as the simulator requires. The dynamics consume them through `B`/`D` (linear-Gaussian) or the `u` argument of a callable transition/observation.
-- `dt` maps integer steps to dsx float times (`obs_times = arange(t_obs) * dt`). It is irrelevant for discrete-time dynamics and is the sampling interval for continuous-time ones.
+- `h` is the current call's `Horizon`. `h.data` selects the mode: training when present and `h.future == 0`, forecasting when `h.future > 0`, in-sample predictive when `None`. The observations themselves come from `y`.
```
Comment (mattlevine22) [inline DYNESTYX_INTEGRATION_DESIGN.md]: Can it be more clear how Horizon encodes the forecasted times?

I don't want to open too big a can of worms here, but if the observations can come at "any" time-grid, then I'd guess so can the forecasts :).

File: DYNESTYX_INTEGRATION_DESIGN.md [hunk 4]
```diff
@@ -1,411 +1,539 @@
...
+`LTI_discrete` infers `control_dim` from `D` when `B` is `None` since dynestyx#361; on dynestyx 0.5.0 the same model needs the explicit `DynamicalModel(..., control_dim=k)` constructor (Section 10, item 7).
+
+An observation-Markov model with the exact-observation idiom, an AR(1) whose state is the observation itself:
+
+```python
+def ar1_dynamics(phi: Array, sigma: Array) -> DynamicalModel:
+    return dsx.DynamicalModel(
+        initial_condition=dist.Normal(0.0, sigma / jnp.sqrt(1 - phi**2)).expand([1]).to_event(1),
+        state_evolution=lambda x, u, t_now, t_next: dist.Normal(phi * x, sigma).to_event(1),
+        observation_model=dsx.DiracIdentityObservation(),
```
Comment (mattlevine22) [inline DYNESTYX_INTEGRATION_DESIGN.md:459]: yes, nice!

File: DYNESTYX_INTEGRATION_DESIGN.md [hunk 5]
```diff
@@ -1,411 +1,539 @@
...
+    )
 ```
 
-`LTI_discrete` infers `control_dim` from `B` alone, so a model whose controls enter only through `D` must use the explicit `DynamicalModel` constructor with `control_dim` (Appendix A.4; also Section 17).
+Under a `LatentPathBuilder` this model has no latent site at all (the path is the data), its factor is the exact likelihood $\sum_t \log p(y_t \mid y_{t-1}, \theta)$, its rollout is an AR(1) forecast and its in-sample predictive returns the data (A.9).
```
Comment (mattlevine22) [inline DYNESTYX_INTEGRATION_DESIGN.md:463]: exactly!

**Review by mattlevine22 — COMMENTED** [review]
All makes sense to me. Added some comments; I'll take a look at the notebooks as well before "Approving" here.

File: DYNESTYX_INTEGRATION_DESIGN.md (hunk 2 above)
Comment (mattlevine22) [reply DYNESTYX_INTEGRATION_DESIGN.md]: After looking at the notebook, it seems you do want the Smoother! Great!

**Comment by mattlevine22** [comment]
@juanitorduz We bumped Dynestyx's version so that you don't need to use a branch now (`v0.5.1`)

File: DYNESTYX_INTEGRATION_DESIGN.md (hunk 1 above)
Comment (juanitorduz) [reply DYNESTYX_INTEGRATION_DESIGN.md]: Thanks! adeed come comments in https://github.com/juanitorduz/numpyro_forecast/pull/125/commits/c41b251d6d13843f636ef2374a4ec7dc1984232b :)

File: DYNESTYX_INTEGRATION_DESIGN.md (hunk 3 above)
Comment (juanitorduz) [reply DYNESTYX_INTEGRATION_DESIGN.md]: added a clarification :)
